src/checkpoint/dispatch/dispatch.impl.h

```diff
@@ -60,22 +65,75 @@ inline Serializer& operator|(Serializer& s, T& target) {
 namespace checkpoint { namespace dispatch {
 
 template <typename T, typename TraverserT>
-TraverserT& Traverse::with(T& target, TraverserT& t, SerialSizeType len) {
+TraverserT& withTypeIdx(TraverserT& t) {
+  using CleanT = typename CleanType<typeregistry::DecodedIndex>::CleanT;
+  using DispatchType =
+    typename TraverserT::template DispatcherType<TraverserT, CleanT>;
+
+  auto const thisTypeIdx = typeregistry::getTypeIdx<T>();
+  SerializerDispatch<TraverserT, CleanT, DispatchType> ap;
+  constexpr SerialSizeType len = 1;
+  if (t.isPacking() || t.isSizing()) {
+    auto val = cleanType(&thisTypeIdx);
+    ap(t, val, len);
+  } else if (t.isUnpacking()) {
+    typeregistry::DecodedIndex serTypeIdx = 0;
+    auto val = cleanType(&serTypeIdx);
+    ap(t, val, len);
+
+    if (
+      typeregistry::validateIndex(serTypeIdx) == false ||
+      thisTypeIdx != serTypeIdx
+    ) {
+      auto err = std::string("Unpacking wrong type, got=") +
+        typeregistry::getTypeNameForIdx(thisTypeIdx) +
+        " idx=" + std::to_string(thisTypeIdx) +
+        ", expected=" + typeregistry::getTypeNameForIdx(serTypeIdx) +
+        " idx=" + std::to_string(serTypeIdx);
+      throw std::runtime_error(err);
+    }
+  }
 
+  return t;
+}
+
+template <typename T, typename TraverserT>
+TraverserT& Traverse::with(T& target, TraverserT& t, SerialSizeType len) {
   using CleanT = typename CleanType<T>::CleanT;
-  using DispatchType = typename TraverserT::template DispatcherType<TraverserT, CleanT>;
+  using DispatchType =
+    typename TraverserT::template DispatcherType<TraverserT, CleanT>;
 
-  auto val = cleanType(&target);
+  #if defined(SERIALIZATION_ERROR_CHECKING)
+  withTypeIdx<CleanT>(t);
+  #endif
 
+  auto val = cleanType(&target);
   SerializerDispatch<TraverserT, CleanT, DispatchType> ap;
+
+  #if defined(SERIALIZATION_ERROR_CHECKING)
+  try {
+    ap(t, val, len);
+  } catch (std::runtime_error& err) {
```

Comment:
If the code is going to throw and catch for this, it should probably use a distinct exception type, so that we don't inadvertently catch some other unrelated exception that happens to arise e.g. in a user-written `serialize()` method.

Reply:
With a custom exception type, the `path` lines should also include an incrementing number indicating the effective call depth.

Reply:
I completely agree, thanks for spotting this!

Reply:
@PhilMiller done:

```
Unpacking wrong type, got=int (idx=3724543661), expected= (idx=1231123)
#0 int
#1 checkpoint::tests::unit::AA
#2 checkpoint::tests::unit::BB
#3 checkpoint::tests::unit::CC
#4 checkpoint::tests::unit::DD
#5 checkpoint::tests::unit::EE
```